keras/src/callbacks/memory_usage_callback.py

```diff
@@ -0,0 +1,192 @@
+import os
+import warnings
+
+from keras.src import backend as K
+from keras.src.api_export import keras_export
+from keras.src.callbacks.callback import Callback
+
+# Attempt to import psutil for CPU memory monitoring
+try:
+    import psutil
+except ImportError:
+    psutil = None
+
+
+@keras_export("keras.callbacks.MemoryUsageCallback")
+class MemoryUsageCallback(Callback):
+    """Monitors and logs memory usage
+    (CPU + optional GPU/TPU/OpenVINO) during training.
+
+    This callback measures:
+
+      - **CPU**: via psutil.Process().memory_info().rss
+      - **GPU/TPU**: via backend‐specific APIs
+        (TensorFlow, PyTorch, JAX)
+
+    Logs are printed to stdout at the start/end of each epoch and,
+    if `log_every_batch=True`, after every batch.
+    If `tensorboard_log_dir` is provided, scalars are also written
+    via `tf.summary` (TensorBoard).
+
+    Args:
+        monitor_gpu (bool): If True, attempt to measure accelerator memory.
+        log_every_batch (bool): If True, also log after each batch.
+        tensorboard_log_dir (str|None): Directory for TensorBoard logs;
+            if None, no TF summary writer is created.
+
+    Raises:
+        ImportError: If `psutil` is not installed (required for CPU logging).
+    """
+
+    def __init__(
+        self,
+        monitor_gpu=True,
+        log_every_batch=False,
+        tensorboard_log_dir=None,
+    ):
+        super().__init__()
+        if psutil is None:
+            raise ImportError(
+                "MemoryUsageCallback requires the 'psutil' library. "
+                "Install via `pip install psutil`."
+            )
+        self.monitor_gpu = monitor_gpu
+        self.log_every_batch = log_every_batch
+        self._proc = psutil.Process()
+        self._step_counter = 0
+        self._writer = None
+
+        if tensorboard_log_dir:
+            try:
+                import tensorflow as tf
+
+                logdir = os.path.expanduser(tensorboard_log_dir)
+                self._writer = tf.summary.create_file_writer(logdir)
+                print(f"MemoryUsageCallback: TensorBoard logs → {logdir}")
+            except Exception as e:
+                warnings.warn(
+                    f"Could not initialize TensorBoard writer: {e}",
+                    RuntimeWarning,
+                )
+                self._writer = None
+
+    def on_train_begin(self, logs=None):
+        self._step_counter = 0
+
+    def on_epoch_begin(self, epoch, logs=None):
+        self._log_epoch("start", epoch)
+
+    def on_epoch_end(self, epoch, logs=None):
+        self._log_epoch("end", epoch, offset=1)
```

Comment:
from the colab output I am observing that Epoch end is not logged when log_every_batch is False